prometheus-operator: Move the CRDs to its own chart

surajssd opened the pull request:
This commit moves the prometheus-operator CRDs to its own chart.

This removes the cyclic dependency of components needing the
prometheus-operator CRDs installed before they can enable the
monitoring. This is especially the case with storage applications,
prometheus-operator needs storage and storage can only enable monitoring
after prometheus-operator is installed.

Fixes #750 
Fixes #557

invidian reviewed (changes requested):
What about https://github.com/kinvolk/lokomotive/pull/567#issuecomment-651558677? I still think those CRDs do not belong to control-plane in any way.

EDIT: And entire #567 actually, which also provides solution for those issues.

surajssd commented:
> The difference though is, that PodSecurityPolicy and NetworkPolicy are native resources, consumed by Kubernetes itself. ServiceMonitor is fully 3rd party resource. As a user, if I don't plan to install monitoring, I'd also expect to not find the CRDs installed.

We can add a knob to disable such installation, if user does not want the monitoring solution shipped by Lokomotive.

As of now we ship multiple monitoring resources that are not part of Prometheus Operator but they reside in control plane charts like the services (endpoints) that expose the monitoring ports of the control plane.

This PR enables us to monitor Calico, which we don't do currently.

invidian commented:
>We can add a knob to disable such installation, if user does not want the monitoring solution shipped by Lokomotive.

That knob would be on controlplane level, so where it doesn't belong. 

>As of now we ship multiple monitoring resources that are not part of Prometheus Operator but they reside in control plane charts like the services (endpoints) that expose the monitoring ports of the control plane.

In regular installations, kube-prometheus-stack delivers all controlplane services which are required for monitoring. Rather than going down with doing something the wrong way, I would rather fix remaining shortcomings.

I don't see much of a difference between this PR and #567. Both are workarounds for the deeper issues.

Commit:
prometheus-operator: Move the CRDs to separate chart

This commit moves the prometheus-operator CRDs to its own chart.

This removes the cyclic dependency of components needing the
prometheus-operator CRDs installed before they can enable the
monitoring. This is especially the case with storage applications,
prometheus-operator needs storage and storage can only enable monitoring
after prometheus-operator is installed.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
EM: Add variable install_prometheus_operator_crds

User can enable/disable the installation of the Prometheus Operator CRDs
using the parameter `install_prometheus_operator_crds`.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
AWS: Add variable install_prometheus_operator_crds

User can enable/disable the installation of the Prometheus Operator CRDs
using the parameter `install_prometheus_operator_crds`.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
codespell: Skip prometheus-operator CRDs

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

knrt10 commented on assets/charts/control-plane/prometheus-operator-crds/Chart.yaml:
```diff
@@ -0,0 +1,6 @@
+description: A Helm chart for Prometheus Operator CRDS
+type: application
+version: 0.1.0
+appVersion: "0.1.0"
```
It would be nice if we follow consistency. 

```suggestion
appVersion: 0.1.0
```

knrt10 commented on assets/charts/control-plane/prometheus-operator-crds/Chart.yaml:
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: prometheus-operator-crds
+description: A Helm chart for Prometheus Operator CRDS
```
We should follow per k8s documentation.

```suggestion
description: A Helm chart for Prometheus Operator CRDs.
```

knrt10 reviewed (changes requested):
Small NITs, apart from CI failing, LGTM.

surajssd commented:
I am not sure why `codespell` keeps failing, it does not fail locally.

knrt10 commented:
> I am not sure why codespell keeps failing, it does not fail locally.

@surajssd It because when running locally codespell uses the file `.codespell.skip`, but for CI you have add `./assets/charts/control-plane/prometheus-operator-crds/templates` also to `ci.yaml` file as currently their github action does not support file format.

surajssd commented:
> `install_prometheus_operator_crds` for the AKS platform too, as some components support monitoring

Prom Op CRDs is a control-plane only chart. Right now we don't have a way to install charts that are designated as control-plane and installed on managed platforms. There is not a way to add this component.
